pkg/utils/rbac/access_review.go

```diff
@@ -0,0 +1,142 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package rbac
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	authorizationapi "k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/kubernetes"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	ServiceAccountUsernamePrefix = "system:serviceaccount:"
+)
+
+var log = logf.Log.WithName("access-review")
+
+type AccessReviewer interface {
+	// AccessAllowed checks that the given ServiceAccount is allowed to update an other object.
+	AccessAllowed(serviceAccount string, sourceNamespace string, object runtime.Object) (bool, error)
+}
+
+type subjectAccessReviewer struct {
+	client kubernetes.Interface
+}
+
+var _ AccessReviewer = &subjectAccessReviewer{}
+
+func NewSubjectAccessReviewer(client kubernetes.Interface) AccessReviewer {
+	return &subjectAccessReviewer{
+		client: client,
+	}
+}
+
+func NewPermissiveAccessReviewer() AccessReviewer {
+	return &permissiveAccessReviewer{}
+}
+
+func (s *subjectAccessReviewer) AccessAllowed(serviceAccount string, sourceNamespace string, object runtime.Object) (bool, error) {
+	metaObject, err := meta.Accessor(object)
+	if err != nil {
+		return false, nil
+	}
+	// For convenience we still allow association between objects in a same namespace
+	if sourceNamespace == metaObject.GetNamespace() {
+		return true, nil
+	}
+
+	if len(serviceAccount) == 0 {
+		serviceAccount = "default"
+	}
+
+	allErrs := field.ErrorList{}
+	// This validation could be done in other places but it is important to be sure that it is done before any access review.
+	for _, msg := range validation.IsDNS1123Subdomain(serviceAccount) {
+		allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInvalid, Field: "serviceAccount", BadValue: serviceAccount, Detail: msg})
+	}
+	if len(allErrs) > 0 {
+		return false, allErrs.ToAggregate()
+	}
+
+	kind := object.GetObjectKind().GroupVersionKind().Kind
+	plural, err := toPlural(kind)
+	if err != nil {
+		return false, nil
+	}
+
+	sar := &authorizationapi.SubjectAccessReview{
+		Spec: authorizationapi.SubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Namespace: metaObject.GetNamespace(),
+				Verb:      "get",
+				Resource:  plural,
+				Group:     strings.ToLower(object.GetObjectKind().GroupVersionKind().Group),
+				Version:   strings.ToLower(object.GetObjectKind().GroupVersionKind().Version),
+				Name:      metaObject.GetName(),
+			},
+			User: ServiceAccountUsernamePrefix + sourceNamespace + ":" + serviceAccount,
+		},
+	}
```

Comment:
Can we extract this to its own function and unit test it? Just because of the string manipulation.